Public lhlawson cpugpuuncore fixed freq

scoumeri opened the pull request:
- Fixes #2008 
- Fixes #2009  change request from github issues.

[Brief description of the change]
Added fixed frequency agent to set accelerator, cpu, and uncore frequencies.  This is being set as a draft PR as the unit tests are not completed yet.

lhlawson committed:
First draft of FixedFrequencyAgent for CPU, GPU, and Uncore experiments

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Added Frequency Max alias and started documentation

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

scoumeri committed:
Added documentation for fixed frequency agent

Modified fixed frequency agent initial conditions

Signed-off-by: Sari Coumeri <sari.coumeri@intel.com>

lhlawson commented on src/FixedFrequencyAgent.hpp:
```diff
@@ -0,0 +1,118 @@
+                M_NUM_SAMPLE
+            };
+
+            PlatformIO &m_platform_io;
+            const PlatformTopo &m_platform_topo;
+
+            geopm_time_s m_last_wait;
+            const double M_WAIT_SEC;
+
+            bool m_do_write_batch;
+            bool m_is_adjust_initialized;
```
Change ordering to match that used in initialization

scoumeri replied:
Fixed

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,309 @@
+    {
+        return {"ACCELERATOR_FREQUENCY",
+                "CORE_FREQUENCY",
+                "UNCORE_FREQUENCY",
```
Do we want to provide the user the ability to input a min & max uncore frequency?

scoumeri replied:
Updated to have UNCORE_MIN_FREQUENCY and UNCORE_MAX_FREQUENCY

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,309 @@
+        }
+
+        if (!m_is_adjust_initialized) {
+            double cpu_init_freq = m_platform_io.read_signal("CPU_FREQUENCY_CONTROL", GEOPM_DOMAIN_BOARD, 0);
```
We're handling the initial CPU & Uncore frequency control here, but the GPU control above (through guaranteeing a policy value).  It may be better to just explicitly handle all initial writes here.

scoumeri replied:
Fixed

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,309 @@
+        , m_do_write_batch(false)
+        , m_is_adjust_initialized(false)
+        , m_signal_available({{"FREQUENCY_ACCELERATOR", {}},
+                              {"NVML::UTILIZATION_ACCELERATOR", {}},
+                              {"NVML::UTILIZATION_MEMORY", {}},
```
```suggestion
```
These signals aren't needed for this agent

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,309 @@
+        , m_do_write_batch(false)
+        , m_is_adjust_initialized(false)
+        , m_signal_available({{"FREQUENCY_ACCELERATOR", {}},
+                              {"NVML::UTILIZATION_ACCELERATOR", {}},
+                              {"NVML::UTILIZATION_MEMORY", {}},
```
```suggestion
```
These signals aren't required for this agent.  The power and energy consumption aren't actually required either, but are likely to be of interest to GPU fixed frequency runs so we may wish to keep them in the trace list.

scoumeri replied:
Fixed

scoumeri committed:
Updated based on PR feedback

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
Added fixed_frequency to agent names in TestAgent.py

Signed-off-by: slcoumer <sari.coumeri@intel.com>

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,306 @@
+        if (!m_is_adjust_initialized) {
+            double accel_init_freq = m_platform_io.read_signal("NVML::FREQUENCY_MAX", GEOPM_DOMAIN_BOARD, 0);
+            freq_ctl_itr = m_control_available.find("FREQUENCY_ACCELERATOR_CONTROL");
+	    m_platform_io.adjust(freq_ctl_itr->second.m_batch_idx, accel_init_freq);
+	    freq_ctl_itr->second.m_last_setting = accel_init_freq;
```
wsp issue

scoumeri replied:
Fixed

scoumeri committed:
Fixed white space issue

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
Modified fixed frequency agent to have a min and max uncore frequency

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
Adding better checks in validate policy for FixedFrequenyAgent

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
Added unit tests

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
Fixed Unit tests

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri commented on ronn/geopm_agent_fixed_frequency.7.ronn:
```diff
@@ -0,0 +1,137 @@
+
+    {"ACCELERATOR_FREQUENCY": 1530000000,
+     "CORE_FREQUENCY": 3100000000,
+     "UNCORE_FREQUENCY": 2000000000}
```
update to have min and max uncore
add sample rate as last parameter: NAN to default to 5ms in code

scoumeri replied:
Done and moved to a .rst file

scoumeri commented on ronn/geopm_agent_fixed_frequency.7.ronn:
```diff
@@ -0,0 +1,137 @@
+
+    $ geopmread NVML::FREQUENCY_MAX board 0
+    1530000000
+
```
Check for name changes GPU_FREQUENCY_MIN_LIMIT etc (Brad's PR)
Update code and documentation

scoumeri replied:
also include #1945 and then change this...

scoumeri replied:
Pulled in latest dev which uses aliases from #1945

scoumeri commented on ronn/geopm_agent_fixed_frequency.7.ronn:
```diff
@@ -0,0 +1,137 @@
+
+    $ geopmread MSR::UNCORE_RATIO_LIMIT:MAX_RATIO board 0
+    2400000000
+
```
Add alias uncore and then update documentation and naming

scoumeri replied:
I have not done this as it requires aliases for uncore not available yet.

scoumeri commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,349 @@
+        // Populate controls for each domain
+        for (auto &sv : m_control_available) {
+            sv.second.m_batch_idx = m_platform_io.push_control(sv.first, GEOPM_DOMAIN_BOARD, 0);
+        }
```
may not need to be a map?

scoumeri replied:
simplified code

scoumeri commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,349 @@
+
+        // set Accelerator frequency control
+        auto freq_ctl_itr = m_control_available.find("FREQUENCY_ACCELERATOR_CONTROL");
+        if (!std::isnan(in_policy[M_POLICY_ACCELERATOR_FREQUENCY])) {
```
change from map to what is in the frequency_map_agent

scoumeri replied:
removed map

scoumeri commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,349 @@
+            freq_ctl_itr = m_control_available.find("FREQUENCY_ACCELERATOR_CONTROL");
+            m_platform_io.adjust(freq_ctl_itr->second.m_batch_idx, accel_init_freq);
+            freq_ctl_itr->second.m_last_setting = accel_init_freq;
+
```
call write control instead of batch and only do all of the code once

scoumeri replied:
don't need to write everything so if a nan don't write anything

scoumeri replied:
simplified code

scoumeri committed:
Rewrote Fixed Frequency Agent to be simpler.  Added additional json parameter to set SAMPLE_PERIOD.

Signed-off-by: slcoumer <sari.coumeri@intel.com>

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -80,24 +74,15 @@ namespace geopm
-        double accel_min_freq = m_platform_io.read_signal("NVML::FREQUENCY_MIN", GEOPM_DOMAIN_BOARD, 0);
-        double accel_max_freq = m_platform_io.read_signal("NVML::FREQUENCY_MAX", GEOPM_DOMAIN_BOARD, 0);
+        double accel_min_freq = m_platform_io.read_signal("NVML::GPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+        double accel_max_freq = m_platform_io.read_signal("NVML::GPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0);
```
If rebasing on top of PR #1945 these can be the aliases: 
GPU_FREQUENCY_MIN_AVAIL
GPU_FREQUENCY_MAX_AVAIL
without the NVML:: prefix.

scoumeri replied:
Pulled in latest dev and uses aliases

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -175,70 +171,38 @@ namespace geopm
+	    
             m_is_adjust_initialized = true;
-            m_do_write_batch = true;
+            //m_do_write_batch = true;
```
This can be deleted.

scoumeri replied:
done

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -338,6 +285,7 @@ namespace geopm
                 "CORE_FREQUENCY",
                 "UNCORE_MIN_FREQUENCY",
                 "UNCORE_MAX_FREQUENCY",
+		"SAMPLE_PERIOD",
```
wsp

scoumeri replied:
done

lhlawson commented on src/FixedFrequencyAgent.hpp:
```diff
@@ -82,6 +82,7 @@ namespace geopm
                 M_POLICY_CPU_FREQUENCY,
                 M_POLICY_UNCORE_MIN_FREQUENCY,
                 M_POLICY_UNCORE_MAX_FREQUENCY,
+		M_POLICY_SAMPLE_PERIOD,
```
wsp

lhlawson commented on src/FixedFrequencyAgent.hpp:
```diff
@@ -93,26 +94,10 @@ namespace geopm
-            const double M_WAIT_SEC;
-
+            double M_WAIT_SEC;
             bool m_do_write_batch;
```
If this is unused it should be removed.

scoumeri replied:
this is still being used, I needed it to not be a constant as it is set by the SAMPLE_PERIOD in the policy.

lhlawson commented on test/FixedFrequencyAgentTest.cpp:
```diff
@@ -71,13 +71,15 @@ class FixedFrequencyAgentTest : public :: testing :: Test
            UNCORE_MIN_CTL_IDX,
-           UNCORE_MAX_CTL_IDX
+           UNCORE_MAX_CTL_IDX,
+	   SAMPLE_PERIOD_IDX
```
wsp

lhlawson commented on test/FixedFrequencyAgentTest.cpp:
```diff
@@ -125,9 +127,9 @@ void FixedFrequencyAgentTest::SetUp()
     ON_CALL(*m_platform_io, control_domain_type("FREQUENCY_ACCELERATOR_CONTROL"))
         .WillByDefault(Return(GEOPM_DOMAIN_BOARD_ACCELERATOR));
-    ON_CALL(*m_platform_io, read_signal("NVML::FREQUENCY_MIN", GEOPM_DOMAIN_BOARD, 0))
+    ON_CALL(*m_platform_io, read_signal("NVML::GPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0))
```
Same note on GPU_FREQUENCY_MIN/MAX_AVAIL vs NVML::GPU_FREQUENCY_MIN/MAX_AVAIL

scoumeri replied:
Latest dev pulled in to use aliases.

scoumeri committed:
Updating to latest geopm dev

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
Added fixed_frequency .rst file to MANIFEST_EXEMPT

scoumeri committed:
Updated for Sample Period in FixedFrequency Agent

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
Added fixed frequency doc to Makefile.mk

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
Fixed Issue with fixed_frequency.7.rst file

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
Changed all names from ACCELERATOR to GPU

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
Updated license headers to 2022

Signed-off-by: slcoumer <sari.coumeri@intel.com>

lhlawson commented on service/docs/source/geopm_agent_fixed_frequency.7.rst:
```diff
@@ -0,0 +1,154 @@
+
+.. code-block:: bash
+
+    $ geopmread NVML::FREQUENCY_MIN board 0
```
```suggestion
    $ geopmread GPU_FREQUENCY_MIN_AVAIL board 0
```

scoumeri replied:
fixed

lhlawson commented on service/docs/source/geopm_agent_fixed_frequency.7.rst:
```diff
@@ -0,0 +1,154 @@
+    $ geopmread NVML::FREQUENCY_MIN board 0
+    135000000
+
+    $ geopmread NVML::FREQUENCY_MAX board 0
```
```suggestion
    $ geopmread GPU_FREQUENCY_MAX_AVAIL board 0
```

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,296 @@
+	    double sample_period = in_policy[M_POLICY_SAMPLE_PERIOD];
+	    
+	    if (!std::isnan(sample_period)) {
+		M_WAIT_SEC = sample_period;
```
wsp

scoumeri replied:
fixed

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,296 @@
+        m_do_write_batch = false;
+
+        if (!m_is_adjust_initialized) {
+	    double gpu_freq_request = in_policy[M_POLICY_GPU_FREQUENCY];
```
missing indentation from 175-205

scoumeri replied:
think i fixed?

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,296 @@
+    //If new values have been adjusted, write
+    bool FixedFrequencyAgent::do_write_batch(void) const
+    {
+        return m_do_write_batch;
```
m_do_write_batch is no longer used, could remove var and always return false.

scoumeri replied:
fixed

scoumeri committed:
Updated based on review

Signed-off-by: slcoumer <sari.coumeri@intel.com>

scoumeri committed:
more whitespace issues

Signed-off-by: slcoumer <sari.coumeri@intel.com>

avilcheslopez committed:
Minor clean-up after rebasing onto dev

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Fix FixedFrequencyAgentTest.cpp after rebase

Update the test to use the correct geopm_domain_e enum values, as they
have been renamed.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Minor clean-up after rebasing onto dev

Removed some white space and cruft leftover in documentation files after
resolving merge conflicts.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez commented on service/docs/source/geopm_agent_fixed_frequency.7.rst:
```diff
@@ -0,0 +1,150 @@
+.. role:: raw-html-m2r(raw)
```
File needs to be updated to follow the format being used now in the latest documentation files.

avilcheslopez commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,294 @@
+/*
```
Replace with license identifier

avilcheslopez commented on src/FixedFrequencyAgent.hpp:
```diff
@@ -0,0 +1,102 @@
+/*
```
Replace with license identifier

avilcheslopez commented on test/FixedFrequencyAgentTest.cpp:
```diff
@@ -0,0 +1,256 @@
+/*
```
Replace with license identifier

avilcheslopez commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,294 @@
+    void FixedFrequencyAgent::init_platform_io(void)
+    {
+
+    }
```
Q: why is this method empty? Is it simply stubbed out for now with the intention of completing it in the future?

avilcheslopez commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,294 @@
+    // Validate incoming policy and configure default policy requests.
+    void FixedFrequencyAgent::validate_policy(std::vector<double> &in_policy) const
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
```
GEOPM_DEBUG_ASSERT

avilcheslopez commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,294 @@
+    void FixedFrequencyAgent::validate_policy(std::vector<double> &in_policy) const
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
+        double gpu_min_freq = m_platform_io.read_signal("GPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+        double gpu_max_freq = m_platform_io.read_signal("GPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+        double core_freq_min = m_platform_io.read_signal("FREQUENCY_MIN", GEOPM_DOMAIN_BOARD, 0);
+        double core_freq_max = m_platform_io.read_signal("FREQUENCY_MAX", GEOPM_DOMAIN_BOARD, 0);
```
Need to update signal names

avilcheslopez replied:
See: #2407

avilcheslopez commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,294 @@
+    void FixedFrequencyAgent::split_policy(const std::vector<double>& in_policy,
+                                    std::vector<std::vector<double> >& out_policy)
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
```
GEOPM_DEBUG_ASSERT

avilcheslopez commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,294 @@
+
+    void FixedFrequencyAgent::adjust_platform(const std::vector<double>& in_policy)
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
```
GEOPM_DEBUG_ASSERT

avilcheslopez commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,294 @@
+
+            // set gpu frequency control
+            if (!std::isnan(gpu_freq_request)) {
+                m_platform_io.write_control("GPU_FREQUENCY_CONTROL",GEOPM_DOMAIN_BOARD,0,gpu_freq_request);
```
Adjust control name?

avilcheslopez replied:
Note from Lowren: use GPU_CORE_FREQUENCY_CONTROL
See: #2045

avilcheslopez commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,294 @@
+            }
+
+            // set Uncore frequency controls
+            if (!std::isnan(uncore_min_freq_request)) {
+                m_platform_io.write_control("MSR::UNCORE_RATIO_LIMIT:MIN_RATIO",GEOPM_DOMAIN_BOARD,0,uncore_min_freq_request);
+            }
+
+            if (!std::isnan(uncore_max_freq_request)) {
+                m_platform_io.write_control("MSR::UNCORE_RATIO_LIMIT:MIN_RATIO",GEOPM_DOMAIN_BOARD,0,uncore_min_freq_request); 
+            }
```
Use high-level aliases?

avilcheslopez replied:
Note from Lowren: changes pending to enable doing so: #2420

avilcheslopez reviewed (commented):
Will work on addressing these changes, submitting these comments here just for tracking

avilcheslopez commented on service/docs/source/geopm_agent_fixed_frequency.7.rst:
```diff
@@ -0,0 +1,150 @@
+specified by SAMPLE_PERIOD.  The default is 5ms.
+
+
+EXAMPLES
```
Update signal names in this section

avilcheslopez committed:
Address various review comments

- Documentation formatting
- License headers
- Replace assert with GEOPM_DEBUG_ASSERT
- Minor code clean-up

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

asmaalrawi committed:
Fixed signal names in Fixed frequency agent

bgeltz commented:
Holding off on merging this till post-2.0 because the policy has a sample rate in it.  We need to deal with that in a more general way.

avilcheslopez committed:
Update signal/control names, minor fix

Found a minor issue in the adjust_platform() method of
FixedFrequencyAgent, where it was only setting the CPU uncore min
frequency and not the max.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Fix FixedFrequencyAgent documentation, unit tests

Both needed to be updated to reflect renaming changes.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

lhlawson commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,259 @@
+                "CORE_FREQUENCY",
+                "UNCORE_MIN_FREQUENCY",
+                "UNCORE_MAX_FREQUENCY",
+                "SAMPLE_PERIOD",
```
New direction for post-2.0 is to not provide sample_period as part of the policy, but to do so via the geopmlaunch command.   We should remove SAMPLE_PERIOD from all post-2.0 agent.s

avilcheslopez replied:
Got it. Thanks for the note. So this is the whole deal with the sample period thing that was mentioned earlier when we reviewed the PR.

avilcheslopez commented on src/FixedFrequencyAgent.cpp:
```diff
@@ -0,0 +1,259 @@
+			        std::to_string(in_policy[M_POLICY_CPU_FREQUENCY]) + ".",
+			        GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+
+	  }
```
Nit: fix spacing/alignment
